main.py

```diff
@@ -62,6 +62,7 @@ def setup(self):
         self.topo_controller = self.get_topo_controller()
         Link.register_status_func(f"{self.napp_id}_link_up_timer",
                                   self.link_status_hook_link_up_timer)
+        self.link_status_change_cache = {}
```

Comment:
Two minor things to improve code readability and its purpose:

1 - Nuke the `setdefault` usages and replace here `{}` with `defaultdict(lambda: {})`, and then add a description saying that this is a dict to keep track of a link id to their states, type hint welcome too if not too much cluttered. 
2 - Can you rename it as just `link_status_change`? or something else? Let's keep the word `cache` in our codebase only for actual cache where it's avoid certain recomputations or (expensive) reads. Here it's just storing state.

Reply:
Removed setdefault, replaced with the default dict.